backend/fbdev.c

```diff
@@ -0,0 +1,278 @@
+/*
+ * Twin - A Tiny Window System
+ * Copyright (c) 2024 National Cheng Kung University, Taiwan
+ * All rights reserved.
+ */
+
+#include <fcntl.h>
+#include <linux/fb.h>
+#include <linux/kd.h>
+#include <linux/vt.h>
+#include <stdlib.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <twin.h>
+#include <unistd.h>
+
+#include "linux_input.h"
+#include "twin_backend.h"
+#include "twin_private.h"
+
+#define FBDEV_NAME "FRAMEBUFFER"
+#define FBDEV_DEFAULT "/dev/fb0"
+#define SCREEN(x) ((twin_context_t *) x)->screen
+#define PRIV(x) ((twin_fbdev_t *) ((twin_context_t *) x)->priv)
+
+typedef struct {
+    twin_screen_t *screen;
+
+    /* Linux input system */
+    void *input;
+
+    /* Linux virtual terminal (VT) */
+    int vt_fd;
+    int vt_num;
+    bool vt_active;
+
+    /* Linux framebuffer */
+    int fb_fd;
+    struct fb_var_screeninfo fb_var;
+    struct fb_fix_screeninfo fb_fix;
+    uint16_t cmap[3][256];
+    uint8_t *fb_base;
+    size_t fb_len;
+} twin_fbdev_t;
+
+static void _twin_fbdev_put_span(twin_coord_t left,
+                                 twin_coord_t top,
+                                 twin_coord_t right,
+                                 twin_argb32_t *pixels,
+                                 void *closure)
+{
+    twin_screen_t *screen = SCREEN(closure);
+    twin_fbdev_t *tx = PRIV(closure);
+
+    if (tx->fb_base == MAP_FAILED)
+        return;
+
+    twin_coord_t width = right - left;
+    off_t off = top * screen->width + left;
+    uint32_t *src = pixels;
```

Comment:
Drop the declaration of `src` since placing `pixels` in the parameter list would be more intuitive.